Match file paths with spaces in caller regexp

rwstauner opened the pull request:
When matching only `\S+` I get warnings like this on my ci server:

```
Cannot digest non-existent file: '/tmp/space char20151201-28026-1sux619/some_ruby.rb20151201-28026-g481xg:1:in `<top (required)>''.
Please set `::_cache_digest` of the serializer
if you'd like to cache it.
```

Matching the trailing space after `:\d+:in` wasn't strictly necessary, but seemed like it would help it to be more accurate.

bf4 commented:
duplicate : https://github.com/rails-api/active_model_serializers/pull/1355

rwstauner commented:
rats.  i searched for the "cache" warning but i guess i didn't search for "spaces".

bf4 commented:
you solved it somewhat differently. would you like to merge in changes from the other (mine), or have me add some changes?  I don't have failing tests for the difference between the two regexen

bf4 commented:
(As a maintainer, I'd rather someone else get credit...)

beauby commented:
Great, I was hoping someone would fix the spaces situation. Thanks :+1:

rwstauner commented:
After fiddling with some extra tests I think your pattern is more accurate.

My original thinking was that the string after the file path (an error message) could include a substring match.
But in this case you aren't matching errors, you're matching an explicit caller line, so I assume the ending will only be something like `` `<top (required)>' `` (or `` `<main>' ``),
in which case the greedy regexp is more correct.

An example that passes for non-greedy and fails for greedy:

```ruby
def test_serializer_file_path_with_path_in_string
  # $ ruby =(echo $'def foo(s); raise s; end;\nfoo("file:123:in string")')
  # /tmp/zshr2gfAE:1:in `foo': file:123:in string (RuntimeError)
  #         from /tmp/zshr2gfAE:2:in `<main>'
  path = '/Users/git/ember js/ember-crm-backend/app/serializers/lead_serializer.rb'
  caller_line = "#{path}:1:in `foo': file:123:in string (RuntimeError)"
  assert_equal path, caller_line[ActiveModel::Serializer::CALLER_FILE]
end
```

but as I said, I don't think that's appropriate in this case.

So I adjusted mine, added your test, and added an additional (more convoluted) test to confirm the difference between the two.  Use whichever pull req makes sense.

bf4 commented on test/serializers/cache_test.rb:
```diff
@@ -160,9 +161,23 @@ def test_serializer_file_path_on_windows
+
+      def test_serializer_file_path_with_submatch
+        # The submatch in the path ensures we're using a correctly greedy regexp.
+        path = '/Users/git/ember js/ember:123:in x/app/serializers/lead_serializer.rb'
```
If someone has that in their file path I pity them :smiley_cat:

rwstauner replied:
Yes, quite.

In my case the file path was generated by another process with apparently little sanitation on the naming ;-)

bf4 commented on test/serializers/cache_test.rb:
```diff
@@ -160,9 +161,23 @@ def test_serializer_file_path_on_windows
         contents = "puts 'AMS rocks'!"
-        file = Tempfile.new('some_ruby.rb')
+        dir = Dir.mktmpdir('space char')
+        file = Tempfile.new('some_ruby.rb', dir)
```
good thinking for the integration test here

bf4 commented on lib/active_model/serializer.rb:
```diff
@@ -23,10 +23,11 @@ class Serializer
       (?=      # stop previous match when
         :\d+     # a colon is followed by one or more digits
         :in      # followed by a colon followed by in
+        \x20     # followed by a space
```
Did you notice if the extra space in the lookahead made a difference?

rwstauner replied:
Quite right.  Now that the regexp is greedy that isn't needed.

bf4 commented:
@rwstauner looks good. Could you add a changelog entry as a Fix?

I have a few other comments but nothing big.

I keep thinking there should be a way for the subclass to calculate its own digest outside of the inherited hook, which would be simpler, no?  I think the method would just need to be dynamically defined `define_singleton_method(:cache_digest) { MD5::Digest.hexdigest(__FILE__) }` or something like that, right?

rwstauner commented:
Sounds like a good idea, but we'd have to find another way to get the file from where the class is inheriting (the caller) since the value of `__FILE__` is based on where the token is present.
And `class_eval "..."` isn't any better (`__FILE__` evaluates to `(eval)`).

bf4 commented:
Closes https://github.com/rails-api/active_model_serializers/pull/1355

rwstauner commented:
Thanks!
